### PR TITLE
core: add skipPageLoad to pass config. set for redirect pass

### PR DESCRIPTION
**Pull request by connorjclark**
image

[doc](https://docs.google.com/document/d/1LOtPLphFQGNrF0Di_x0c63FakxQqHZAJ2zFoccM6wMY/edit?resourcekey=0-jQoenPbJyQzXUE8HgVIaPQ#heading=h.t8jrc1m7zvax)

Verified by seeing that http://www.example.com and https://www.example.com still fails the redirect pass, and http://www.paulirish.com and https://www.paulirish.com still passes.

### CODE REVIEW

**Comment by patrickhulce** [comment]
If any future historians are searching for more validation on what `Page.navigate` waits for, it does appear to wait for all HTTP redirects to resolve.

**Example**

```js
const puppeteer = require('puppeteer');

async function go() {
  const browser = await puppeteer.launch({headless: false});
  const page = await browser.newPage();
  const session = await page.target().createCDPSession();
  await session.send('Network.enable');
  session.on('Network.requestWillBeSent', r => console.log('Sent!', r.request.url));
  session.on('Network.responseReceived', r => console.log('Received!', r.response.url));

  console.log('Navigate');
  await session.send('Page.navigate', {url: 'http://optyx.app/download'});
  console.log('Done navigating');

  await browser.close();
}

go()
```

prints

```bash
Navigate
Sent! http://optyx.app/download
Sent! https://us-central1-dwigt-api.cloudfunctions.net/download
Sent! https://cdn.eris.ws/optyx/mac/x64/Optyx-latest.dmg
Received! https://cdn.eris.ws/optyx/mac/x64/Optyx-latest.dmg
Done navigating
```

@brendankenny called this out in the doc, but this does mean that javascript redirects likely won't be respected anymore. Worth calling out in our release notes and/or PR description?

File: lighthouse-core/config/default-config.js [hunk 1]
```diff
@@ -179,6 +179,7 @@ const defaultConfig = {
     gatherers: [
       'http-redirect',
     ],
+    skipPageLoad: true,
```
Comment (adamraine) [inline lighthouse-core/config/default-config.js:182]: What would be the implications of adding this to `offlinePass`?

**Review by adamraine — APPROVED** [review]
Overall LGTM

File: types/config.d.ts [hunk 2]
```diff
@@ -64,6 +64,8 @@ declare global {
          * Ignore means a failure is expected and no action should be taken.
          */
         loadFailureMode?: 'fatal'|'warn'|'ignore';
+        /** Whether the pass should skip the page load. If false, gathering begins as soon as navigation completes. */
```
Comment (brendankenny) [inline types/config.d.ts:67]:
```suggestion
        /** Whether the pass should skip waiting for the page to load. If true, gathering begins as soon as navigation completes. */
```
Is it `s/false/true`? If it is supposed to be false, maybe reword for clarity?

File: types/config.d.ts [hunk 3]
```diff
@@ -64,6 +64,8 @@ declare global {
...
+        skipPageLoad?: boolean;
```
Comment (brendankenny) [inline types/config.d.ts:68]:
```suggestion
        skipWaitingForPageLoad?: boolean;
```
just an idea and a lot longer, but `skipPageLoad` also kind of reads like "skip loading [navigating to] the page", so this might be a bit clearer?

File: lighthouse-core/config/default-config.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/config/default-config.js:182]: I don't think we even need offline pass at all anymore, FR gatherer conversion sort of revealed that

**Review by brendankenny — COMMENTED** [review]
> but this does mean that javascript redirects likely won't be respected anymore. Worth calling out in our release notes and/or PR description?

we never highlighted this in the 7.0 release notes, so maybe it's fine? The PWA smoke tests (like svgomg in `pwa2`) aren't failing, though,  which had to be changed in #11711 when js redirects started going through. Is this going to make `http-redirect` flaky for js redirects?

**Comment by brendankenny** [comment]
If the root cause is #11738 waiting on critical network requests, how many of those requests are js files since that's one of the few remaining file types not blocked by

https://github.com/GoogleChrome/lighthouse/blob/a753350ecc9f5fb147b4028393b45f93e8efac5e/lighthouse-core/config/default-config.js#L178

?

If we're willing to give up js redirects already, does going back to the "block JavaScript" part of option (c) in that doc significantly improve things while we wait for protocol support for an http redirect check?

**Comment by connorjclark** [comment]
Stepping back a bit ... as soon as a request makes it to a `http:` server, there's no guarantee you are reaching the intended host. Shouldn't this check actually be for HSTS? Either we look for the `Strict-Transport-Security` header, or check that host is in the preload list (?), or check that a http redirect happens _because_ of HSTS (is there protocol support for this?).

In the meantime, it seems fine to drop js redirects, which has got to be even less secure than relying on the server to redirect traffic itself :P

**Comment by brendankenny** [comment]
For ideal security, yes, but it's still good to nudge users to https even if the first hop is insecure.

There's also other benefits to the redirect. One of the original purposes of the audit was for PWA user experience, because SWs only work on secure origins, so it was a PWA best practice to make sure users don't accidentally end up on the `http` version of the site and confused why they don't get the install banner/don't see cached data or whatever from their already installed app. That's why it's part of the "PWA Optimized" badge and not installable or security or whatever (and that's why svgomg also proudly lived with not a full PWA checkmark in Lighthouse, because they disagreed with our failing them for their js-only redirect :)

**Comment by brendankenny** [comment]
(I agree with your argument too (e.g. in #3417) and we've lived without js redirect support for this audit for so long that I think it's fine, just making the distinction on the goal of the audit vs `is-on-https`, for instance)

File: lighthouse-core/config/default-config.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/config/default-config.js]: how do you mean? Is this related to what @patrickhulce and @adrianaixba worked out wrt `service-worker` gatherer?

**Comment by brendankenny** [comment]
It's still very suspicious that the svgomg expectations didn't have to change, but I guess we'll just need to keep an eye out for flakiness on that.

I still mostly agree with https://github.com/GoogleChrome/lighthouse/pull/12141#issuecomment-786780496

> If we're willing to give up js redirects already, does going back to the "block JavaScript" part of option (c) in that doc significantly improve things while we wait for protocol support for an http redirect check?

and that a new `waitFor` condition and config option is a lot to trade for 150ms if a default block on js could get us most of the way there (and if js redirects _are_ working with this approach, blocking js on it's own could potentially save even more?), but if @adamraine and @patrickhulce are still good with the trade-off, SGTM

**Comment by patrickhulce** [comment]
Everything in https://github.com/GoogleChrome/lighthouse/pull/12141#issuecomment-819592729 is compelling to me, but I'm also not really exposed to the ramifications of 150ms longer runs. If blocking JS gets us 50% of the way there with time savings, I'd be hard pressed to support adding and maintaining a new dedicated option that no one else is likely to want to use for any other purpose and is only until we get proper protocol support. 

If blocking JS only gets us 25% of the way there or less, I'm fine deferring the tradeoff decision to those more pressured by whatever timeouts prompted this work.

**Comment by connorjclark** [comment]
It's 150ms in the URLs I tested above, but apparently it is ~10% at the 75% percentile, according to LR graphs.

**Comment by brendankenny** [comment]
I thought that was for the entire pass though, not keeping the pass but stopping at `Page.frameNavigated`?

**Comment by patrickhulce** [comment]
> but apparently it is ~10% at the 75% percentile, according to LR graphs.
> I thought that was for the entire pass though, not keeping the pass but stopping at Page.frameNavigated?

A somewhat important distinction. Does the 70% increase at the 75th percentile account for 10% improvement on the entire run? (meaning it was already accounting for ~14% of the run before the change) Or is it that *after* the change the entire redirect pass accounts for 10% of the run?

The fact that we had such a dramatic increase from waiting for more JS suggests blocking JS might be just as effective at accomplishing our goal here, but if a large portion were just waiting for main documents, then I don't imagine anything we do would have much of an impact.
